src/qrc11.asm

```diff
@@ -0,0 +1,333 @@
+
+    ; ------------------------------------------------------------------------
+    ; Interleave message and ecc blocks.
+    ; ------------------------------------------------------------------------
+
+qrc11_interleave:
+    ld hl, qrc11_b1
+    ld de, qrc11_message
+    ld a, 50
+    call qrc11_intl
+    ld hl, qrc11_b2 + 50
+    ldi
+    ld hl, qrc11_b3 + 50
+    ldi
+    ld hl, qrc11_b4 + 50
+    ldi
+    ld hl, qrc11_b5 + 50
+    ldi
+    ld hl, qrc11_b1_ecc
+    ld a, 30
+    call qrc11_intl
+
+    ; ------------------------------------------------------------------------
+    ; Display QR code with checkerboard mask.
+    ; ------------------------------------------------------------------------
+
+    ld hl, qrc11_map
+    ld c, 61
+qrc11_d1:
+        ld b, 61
+qrc11_d2:   push bc
+            ld e, (hl)
+            inc hl
+            ld d, (hl)
+            inc hl
+            ld a, e
+            and 7
+            srl d
+            rr e
+            srl d
+            rr e
+            srl d
+            rr e
+            ld bc, qrc11_message
+            ex de, hl
+            add hl, bc
+            ex de, hl
+            ld b, a
+            ld a, (de)
+            inc b
+qrc11_d3:       rlca
+                djnz qrc11_d3
+            pop bc
+            xor b
+            xor c
+            rrca
+            call nc, qrc11_module
+            djnz qrc11_d2
+        dec c
+        jr nz, qrc11_d1
+    ret
+
+
+    ; ------------------------------------------------------------------------
+    ; Interleave blocks.
+    ; ------------------------------------------------------------------------
+
+qrc11_intl:
+    ldi
+    ld bc, 49 + 30
+    add hl, bc
+    ldi
+    ld c, 50 + 30
+    add hl, bc
+    ldi
+    ld c, 50 + 30
+    add hl, bc
+    ldi
+    ld c, 50 + 30
+    add hl, bc
+    ldi
+    ld bc, qrc11_b1 - qrc11_b5
+    add hl, bc
+    dec a
+    jr nz, qrc11_intl
+    ret
+
+
+    ; ------------------------------------------------------------------------
+    ; Calculate the block ECC.
+    ; ------------------------------------------------------------------------
+
+qrc11_ecc:
+    ; Save block parameters for restoring
+    push hl
+    push de
+    push bc
+    ; Save message block length for later
+    push bc
+    ; Save message block address for later
+    push de
+    ldir
+    ; Zero the 30 bytes where the ECC will be stored.
+    xor a
+    ld b, 30
+qrc11_zero_ecc:
+        ld (de), a
+        inc de
+    djnz qrc11_zero_ecc
+
+    ; HL is the polynomial A.
+    pop hl
+
+    ; IXL is the outer loop counter (i) for the length of A.
+    pop ix
+qrc11_loop_i:
+        ; Save HL as it'll be incremented in the inner loop.
+        push hl
+
+        ; Save A[i] in B to be used inside the inner loop.
+        ld b, (hl)
+
+        ; DE is the polynomial B.
+        ld de, qrc11_ecc_poly
+
+        ; Evaluate the inner loop count limit.
+        ld a, 31
+
+        ; add ixl for dumb assemblers
+	db $dd
+	add a, l
```

Comment:
Could this be a macro, i.e. `ADD_IXL`? I think it would make the code easier to read.

Reply:
Sure. One sec.